### PR TITLE
Dbaas 2751: Adding the stats.py module

### PR DESCRIPTION
**Pull request by sharmanirek**
Implementing: 

-  One Hot Dummy Column custom transformer
- String Index Reconstructor custom transformer (to reverse the above transformation) 
- preprocessing functions 
- Generate pdf and calculate best fitting  distribution to data  
- custom PCA calculations (exposing eigenvalues)
- PCA reconstruction functionality 
- Basic Markov Chain Object

### CODE REVIEW

File: splicemachine/ml/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "/Users/nireksharma/anaconda3/envs/py27/bin/python"
```
Comment (njnygaard) [inline splicemachine/ml/.vscode/settings.json]: This is a local path to a config file. I would recommend putting `.vscode/` in the `.gitignore`.

**Review by njnygaard — CHANGES_REQUESTED** [review]

File: splicemachine/ml/stats.py [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+    """
+
+    String_Columns = []
+    Continuous_Columns = []
+    for _col, _type in df.dtypes: # This is a tuple of (<col name>, data type)
+        if _col in cols_to_exclude:
+            continue
+        if _type == 'string':
+            String_Columns.append(_col)
+        elif _type == 'double' or _type == 'int' or _type == 'float':
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py:156]: Not sure we can assume this. I see int/long types as categorical more often than continuous variables.

File: splicemachine/ml/stats.py [hunk 3]
```diff
@@ -0,0 +1,591 @@
+# Installing dependancies
+import os
+commands_pip = [ # so you can easily rerun :)
+"""curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py""",
+"""python get-pip.py --user""",
+"""export PATH=$PATH:/zeppelin/.local/bin/"""
+]
+
+for cmd_pip in commands_pip:
+    os.system(cmd_pip)
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: These need to be added to requirements.3.txt in dbaas-infrastructure, not installed here.

File: splicemachine/ml/stats.py [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+statsmodels
+"""
+
+print('stats_requirements.txt')
+print('-' * 50)
+print(packages)
+with open('requirements.txt', 'w') as pcks_file:
+    pcks_file.write(packages)
+
+os.system('~/.local/bin/pip install -r stats_requirements.txt --user')
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: see above comment

File: splicemachine/ml/stats.py [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+from pyspark import keyword_only
+
+import pyspark_dist_explore as dist_explore #FIXME-- MUST INSTALL IN REQUIRMENTS
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+
+
+# Custom Transformers
+class OneHotDummies(Transformer, HasInputCol, HasOutputCol):
+    """Transformer to generate dummy columns for categorical variables as a part of a preprocessing pipeline
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: Doc strings (for all functions as well) should match our docs strings in other files (see management.py/context.py)

**Review by Ben-Epstein — CHANGES_REQUESTED** [review]

**Review by Ben-Epstein — APPROVED** [review]

**Review by njnygaard — APPROVED** [review]
Not really my area of expertise.

**Pull request merged**